File: scripts/import-recycling-data.mjs
/**
 * import-recycling-data.mjs
 * Imports master_recycling_directory.csv into the new recyclish-directory Supabase DB.
 * Run: DATABASE_URL="postgres://..." node scripts/import-recycling-data.mjs
 */

import { createReadStream } from 'fs';
import { parse } from 'csv-parse';
import pg from 'pg';
import { fileURLToPath } from 'url';
import path from 'path';
import crypto from 'crypto';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CSV_PATH = path.join(__dirname, '../client/public/data/master_recycling_directory.csv');

const DB_URL = process.env.DATABASE_URL;
if (!DB_URL) {
  console.error('ERROR: DATABASE_URL environment variable is not set.');
  process.exit(1);
}

// ── Category normalization ──────────────────────────────────────────────────
const CATEGORY_MAP = {
  'PlasticRecycling':           'Drop-off Center',
  'Plastic Recycling':          'Drop-off Center',
  'Electronics Recyclers':      'E-Waste',
  'Electronics Recycler':       'E-Waste',
  'E-Waste':                    'E-Waste',
  'HazardousWaste':             'Hazardous Waste',
  'Hazardous Waste':            'Hazardous Waste',
  'CompostingFacility':         'Composting',
  'Composting Facility':        'Composting',
  'Composting':                 'Composting',
  'ScrapMetal':                 'Scrap Metal',
  'Scrap Metal':                'Scrap Metal',
  'RetailTakeBack':             'Retail Take-Back',
  'Retail Take-Back':           'Retail Take-Back',
  'Retail TakeBack':            'Retail Take-Back',
  'CurbsidePickup':             'Curbside Pickup',
  'Curbside Pickup':            'Curbside Pickup',
  'TransferStation':            'Transfer Station',
  'Transfer Station':           'Transfer Station',
  'MaterialRecoveryFacility':   'Material Recovery Facility',
  'Material Recovery Facility': 'Material Recovery Facility',
  'MunicipalRecycling':         'Municipal Recycling',
  'Municipal Recycling':        'Municipal Recycling',
  'Drop-off Center':            'Drop-off Center',
  'DropOff':                    'Drop-off Center',
  'Drop Off':                   'Drop-off Center',
};

const VALID_TYPES = [
  'Drop-off Center', 'Curbside Pickup', 'Retail Take-Back', 'Hazardous Waste',
  'E-Waste', 'Composting', 'Scrap Metal', 'Transfer Station',
  'Material Recovery Facility', 'Municipal Recycling', 'Community Resource', 'Other'
];

function normalizeCategory(raw) {
  if (!raw) return 'Other';
  const mapped = CATEGORY_MAP[raw.trim()];
  if (mapped) return mapped;
  if (VALID_TYPES.includes(raw.trim())) return raw.trim();
  return 'Other';
}

// ── Address parsing ─────────────────────────────────────────────────────────
function parseAddress(fullAddress) {
  // Format: "4601 Riverview Blvd, St. Louis, MO, United States"
  // or:     "123 Main St, Springfield, IL 62701"
  if (!fullAddress) return { street: '', city: '', state: '', zip: '' };
  const parts = fullAddress.split(',').map(p => p.trim());
  const street = parts[0] || '';
  const city = parts[1] || '';
  // State part may contain zip: "MO 62701" or just "MO"
  const stateRaw = parts[2] || '';
  const stateMatch = stateRaw.match(/^([A-Z]{2})\s*(\d{5}(?:-\d{4})?)?/);
  const state = stateMatch ? stateMatch[1] : stateRaw.replace(/[^A-Za-z ]/g, '').trim();
  const zip = stateMatch && stateMatch[2] ? stateMatch[2] : '';
  return { street, city, state, zip };
}

// ── Slug generation ─────────────────────────────────────────────────────────
function slugify(name, address) {
  const base = `${name} ${address}`
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .substring(0, 80);
  const hash = crypto.createHash('md5').update(`${name}|${address}`).digest('hex').substring(0, 6);
  return `${base}-${hash}`;
}

// ── Main import ─────────────────────────────────────────────────────────────
async function main() {
  const pool = new pg.Pool({
    connectionString: DB_URL,
    ssl: { rejectUnauthorized: false },
    max: 5,
  });

  console.log('Connecting to database...');
  const client = await pool.connect();
  console.log('Connected!');

  let inserted = 0;
  let skipped = 0;
  let errors = 0;
  const slugsSeen = new Set();

  const rows = [];

  // Read CSV
  await new Promise((resolve, reject) => {
    createReadStream(CSV_PATH)
      .pipe(parse({ columns: true, skip_empty_lines: true, trim: true }))
      .on('data', row => rows.push(row))
      .on('end', resolve)
      .on('error', reject);
  });

  console.log(`Read ${rows.length} rows from CSV. Starting import...`);

  // Process in batches of 100
  const BATCH = 100;
  for (let i = 0; i < rows.length; i += BATCH) {
    const batch = rows.slice(i, i + BATCH);
    for (const row of batch) {
      try {
        const name = (row['Name'] || '').trim();
        const fullAddress = (row['Address'] || '').trim();
        if (!name || !fullAddress) { skipped++; continue; }

        const { street, city, state, zip } = parseAddress(fullAddress);
        if (!street || !city || !state) { skipped++; continue; }

        const lat = parseFloat(row['Latitude']);
        const lon = parseFloat(row['Longitude']);
        const hasCoords = !isNaN(lat) && !isNaN(lon) && lat !== 0 && lon !== 0;

        const facilityType = normalizeCategory(row['Category'] || row['Facility_Type'] || '');
        const feedstock = (row['Feedstock'] || '').trim();
        const acceptedMaterials = feedstock ? feedstock.split(/[,;]/).map(s => s.trim()).filter(Boolean) : [];

        let slug = slugify(name, fullAddress);
        // Ensure unique slug
        let attempt = 0;
        while (slugsSeen.has(slug)) {
          attempt++;
          slug = slugify(name, fullAddress) + `-${attempt}`;
        }
        slugsSeen.add(slug);

        const locationExpr = hasCoords
          ? `ST_SetSRID(ST_MakePoint($12, $11), 4326)`
          : null;

        const query = `
          INSERT INTO shelters (
            name, slug, address_line1, city, state, zip,
            phone, email, website, hours_of_operation,
            latitude, longitude, location,
            shelter_type, species_served,
            verified, active
          ) VALUES (
            $1, $2, $3, $4, $5, $6,
            $7, $8, $9, $10,
            $11, $12, ${hasCoords ? `ST_SetSRID(ST_MakePoint($12, $11), 4326)` : 'NULL'},
            $13, $14,
            false, true
          )
          ON CONFLICT (slug) DO NOTHING
        `;

        const hoursRaw = (row['Hours'] || '').trim();
        const hoursJson = hoursRaw ? JSON.stringify({ raw: hoursRaw }) : null;

        await client.query(query, [
          name,                                    // $1
          slug,                                    // $2
          street,                                  // $3
          city,                                    // $4
          state,                                   // $5
          zip || '',                               // $6
          (row['Phone'] || '').trim() || null,     // $7
          (row['Email'] || '').trim() || null,     // $8
          (row['Website'] || '').trim() || null,   // $9
          hoursJson,                               // $10
          hasCoords ? lat : null,                  // $11
          hasCoords ? lon : null,                  // $12
          facilityType,                            // $13
          acceptedMaterials,                       // $14
        ]);
        inserted++;
      } catch (err) {
        errors++;
        if (errors <= 5) console.error(`Row error (${row['Name']}):`, err.message);
      }
    }
    process.stdout.write(`\rProgress: ${Math.min(i + BATCH, rows.length)}/${rows.length} rows processed...`);
  }

  client.release();
  await pool.end();

  console.log(`\n\n✅ Import complete!`);
  console.log(`   Inserted: ${inserted}`);
  console.log(`   Skipped:  ${skipped}`);
  console.log(`   Errors:   ${errors}`);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/import_recycling_data.py
#!/usr/bin/env python3.11
"""
Import cleaned recycling CSV data into the recyclish-directory Supabase database
using the Supabase REST API (HTTPS - no direct DB connection needed).
"""
import requests
import csv
import re
import uuid
import time

SUPABASE_URL = "https://ejtigzzdeblbwdpodgih.supabase.co"
SERVICE_ROLE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVqdGlnenpkZWJsYndkcG9kZ2loIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3NDU3MzEyOCwiZXhwIjoyMDkwMTQ5MTI4fQ.1S25EzX8XKCPEueYz_Ed-f_SP88Jwe5zGpPtivBX1VE"

INSERT_HEADERS = {
    "apikey": SERVICE_ROLE_KEY,
    "Authorization": f"Bearer {SERVICE_ROLE_KEY}",
    "Content-Type": "application/json",
    "Prefer": "return=minimal"
}

def slugify(text):
    """Create a URL-safe slug from text."""
    text = str(text).lower().strip()
    text = re.sub(r'[^\w\s-]', '', text)
    text = re.sub(r'[\s_-]+', '-', text)
    text = re.sub(r'^-+|-+$', '', text)
    return text[:80]

def parse_address(address, city, state, zip_code):
    """Parse address fields, handling combined address strings."""
    address = str(address or '').strip()
    city = str(city or '').strip()
    state = str(state or '').strip()
    zip_code = str(zip_code or '').strip()
    
    # If city is missing but address has comma-separated parts, try to parse
    if not city and ',' in address:
        parts = [p.strip() for p in address.split(',')]
        if len(parts) >= 3:
            address = parts[0]
            city = parts[1]
            state_zip = parts[2].strip().split()
            state = state_zip[0] if state_zip else state
            zip_code = state_zip[1] if len(state_zip) > 1 else zip_code
        elif len(parts) == 2:
            address = parts[0]
            city = parts[1]
    
    return address, city, state, zip_code

def import_csv():
    """Import the cleaned CSV data into the shelters table."""
    csv_path = "/home/ubuntu/recyclish-review/client/public/data/master_recycling_directory.csv"
    
    print(f"Reading CSV from {csv_path}...")
    
    rows = []
    seen_slugs = set()
    skipped = 0
    
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        headers = reader.fieldnames
        print(f"CSV columns: {headers}")
        
        for row in reader:
            # Try multiple possible column name variants
            name = (row.get('Name') or row.get('name') or row.get('Facility Name') or '').strip()
            if not name:
                skipped += 1
                continue
            
            raw_address = (row.get('Address') or row.get('address') or row.get('Street') or '').strip()
            raw_city = (row.get('City') or row.get('city') or '').strip()
            raw_state = (row.get('State') or row.get('state') or '').strip()
            raw_zip = (row.get('Zip') or row.get('zip') or row.get('ZipCode') or row.get('Postal Code') or '').strip()
            
            address, city, state, zip_code = parse_address(raw_address, raw_city, raw_state, raw_zip)
            
            if not city or not state:
                skipped += 1
                continue
            
            # Skip non-recycling entries
            category = (row.get('Category') or row.get('category') or row.get('Type') or '').strip()
            if 'animal' in category.lower() or 'shelter' in category.lower():
                skipped += 1
                continue
            
            # Generate unique slug
            base_slug = slugify(f"{name}-{city}-{state}")
            slug = base_slug
            counter = 1
            while slug in seen_slugs:
                slug = f"{base_slug}-{counter}"
                counter += 1
            seen_slugs.add(slug)
            
            # Parse coordinates
            lat_raw = row.get('Latitude') or row.get('latitude') or row.get('Lat') or ''
            lon_raw = row.get('Longitude') or row.get('longitude') or row.get('Lon') or ''
            
            try:
                lat = float(lat_raw) if lat_raw and str(lat_raw).strip() not in ('', '0', '0.0') else None
                lon = float(lon_raw) if lon_raw and str(lon_raw).strip() not in ('', '0', '0.0') else None
            except (ValueError, TypeError):
                lat = None
                lon = None
            
            feedstock = (row.get('Feedstock') or row.get('feedstock') or row.get('Materials') or row.get('Accepted Materials') or '').strip()
            phone = (row.get('Phone') or row.get('phone') or row.get('Phone Number') or '').strip()
            website = (row.get('Website') or row.get('website') or row.get('URL') or '').strip()
            
            # Parse materials into array
            materials = [m.strip() for m in feedstock.split(',') if m.strip()] if feedstock else []
            
            rows.append({
                "id": str(uuid.uuid4()),
                "name": name,
                "slug": slug,
                "address_line1": address if address else f"{city}, {state}",
                "city": city,
                "state": state,
                "zip": zip_code if zip_code else "00000",
                "latitude": lat,
                "longitude": lon,
                "shelter_type": category if category else None,
                "species_served": materials,
                "phone": phone if phone else None,
                "website": website if website else None,
                "verified": False,
                "active": True,
                "is_no_kill": True,
            })
    
    print(f"Prepared {len(rows)} rows for import. Skipped {skipped} invalid rows.")
    
    # Insert in batches of 200 via Supabase REST API
    batch_size = 200
    total_inserted = 0
    errors = 0
    
    for i in range(0, len(rows), batch_size):
        batch = rows[i:i+batch_size]
        resp = requests.post(
            f"{SUPABASE_URL}/rest/v1/shelters",
            headers=INSERT_HEADERS,
            json=batch,
            timeout=60
        )
        if resp.status_code in (200, 201):
            total_inserted += len(batch)
            print(f"  Batch {i//batch_size + 1}: ✅ {len(batch)} rows ({total_inserted} total)")
        else:
            errors += 1
            print(f"  Batch {i//batch_size + 1}: ❌ {resp.status_code}: {resp.text[:300]}")
        time.sleep(0.3)
    
    print(f"\nImport complete: {total_inserted}/{len(rows)} rows inserted. {errors} batch errors.")
    return total_inserted

if __name__ == "__main__":
    print("=== recyclish-directory CSV Import ===\n")
    count = import_csv()
    print(f"\n✅ Done! {count} recycling facilities now in the database.")
